src/Arcus.WebApi.Security/Authentication/SharedAccessKey/SharedAccessKeyAuthenticationAttribute.cs

```diff
@@ -25,14 +26,27 @@ public class SharedAccessKeyAuthenticationAttribute : TypeFilterAttribute
         /// <param name="headerName">The name of the request header which value must match the stored secret.</param>
         /// <param name="queryParameterName">The name of the query parameter which value must match the stored secret.</param>
         /// <param name="secretName">The name of the secret that's being retrieved using the <see cref="ISecretProvider.GetRawSecretAsync"/> call.</param>
-        /// <exception cref="ArgumentException">When the <paramref name="headerName"/> is <c>null</c> or blank.</exception>
-        /// <exception cref="ArgumentException">When the <paramref name="secretName"/> is <c>null</c> or blank.</exception>
-        public SharedAccessKeyAuthenticationAttribute(string secretName, string headerName = null, string queryParameterName = null) : base(typeof(SharedAccessKeyAuthenticationFilter))
+        /// <exception cref="ArgumentException">Thrown when the <paramref name="secretName"/> is blank.</exception>
+        /// <exception cref="ArgumentException">Thrown when the <paramref name="headerName"/> and the <paramref name="queryParameterName"/> are blank.</exception>
+        public SharedAccessKeyAuthenticationAttribute(string secretName, string headerName = null, string queryParameterName = null) 
+            : base(typeof(SharedAccessKeyAuthenticationFilter))
         {
-            Guard.For<ArgumentException>(() => String.IsNullOrWhiteSpace(headerName) && String.IsNullOrWhiteSpace(queryParameterName), "Either header name or query parameter name must be supplied.");
             Guard.NotNullOrWhitespace(secretName, nameof(secretName), "Secret name cannot be blank");
+            Guard.For<ArgumentException>(
+                () => String.IsNullOrWhiteSpace(headerName) && String.IsNullOrWhiteSpace(queryParameterName), 
+                "Requires either a non-blank header name or query parameter name");
 
-            Arguments = new object[] { headerName?? String.Empty, queryParameterName?? String.Empty, secretName };
+            const bool emitSecurityEvents = false;
+            Arguments = new object[] { headerName?? String.Empty, queryParameterName?? String.Empty, secretName, emitSecurityEvents };
+        }
+
+        /// <summary>
+        /// Gets or sets the flag indicating whether or not the shared access key authentication should emit security events during the authentication process of the request.
+        /// </summary>
+        public bool EmitSecurityEvents
+        {
+            get => Arguments?.LastOrDefault() is bool emitSecurityEvents && emitSecurityEvents;
```

Comment:
Cna't we use an object here? What if we add a new argument, then this will be broken, no?

Reply:
Yes, but that was already the case before these changes. The security attributes already are fixed with their security filter counterparts. The attributes reference towards the type of the filter and send a series of arguments with it to create an filter instance. So, it would already break but I guess in this case it's okay because they 'work together'. They require each other.
Adding this extra property to alter the arguments as an 'afterthought' doesn't make it weaker. Our tests will make sure that things still keeps working if we do change things.

Reply:
> Cna't we use an object here? What if we add a new argument, then this will be broken, no?

Not sure what you mean with an 'object', though? We can't alter the filter (which gets created based on the input from the attribute) with a property. So, we have to pass it as an argument.

Reply:
After consideration, I removed this direct approach and used a separate options instance that will be altered via the property. Better and safer. Thanks.